gateway/middleware_retrieval_timeout.go

```diff
@@ -0,0 +1,182 @@
+// If no data is written within the specified duration, the connection is
+// terminated with a 504 Gateway Timeout.
+//
+// Parameters:
+//   - handler: The HTTP handler to wrap with retrieval timeout
+//   - timeout: Maximum duration between writes (0 disables timeout)
+//   - c: Optional configuration for controlling error page rendering (can be nil)
+func withRetrievalTimeout(handler http.Handler, timeout time.Duration, c *Config, metrics *middlewareMetrics) http.Handler {
+	if timeout <= 0 {
+		return handler
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Create channels for coordination
+		handlerDone := make(chan struct{})
+		timeoutChan := make(chan struct{})
+
+		// Create a custom response writer that tracks writes
+		tw := &timeoutWriter{
+			ResponseWriter: w,
+			timeout:        timeout,
+			timer:          time.NewTimer(timeout),
+			done:           handlerDone,
+			timeoutSignal:  timeoutChan,
+			request:        r,
+			config:         c,
+		}
+
+		// Start the timeout monitoring goroutine
+		go func() {
+			for {
+				select {
+				case <-tw.timer.C:
+					tw.mu.Lock()
+					if !tw.timedOut && !tw.handlerComplete {
+						tw.timedOut = true
+
+						if !tw.wroteHeader {
+							// Headers not sent yet, we can send 504
+							metrics.recordTimeout(http.StatusGatewayTimeout, false)
+							message := "Unable to retrieve content within timeout period"
+							writeErrorResponse(tw.ResponseWriter, tw.request, tw.config, http.StatusGatewayTimeout, message)
+						} else {
+							// Headers already sent, response is being truncated
+							statusCode := tw.headerCode
+							if statusCode == 0 {
+								statusCode = http.StatusOK
+							}
+							metrics.recordTimeout(statusCode, true)
+
+							// Try to write truncation message (best effort)
+							tw.ResponseWriter.Write([]byte(truncationMessage))
+
+							// Try to hijack and force connection reset
+							if hijacker, ok := tw.ResponseWriter.(http.Hijacker); ok {
+								conn, _, err := hijacker.Hijack()
+								if err == nil {
+									tw.hijacked = true
+									// Force TCP RST instead of graceful close
+									if tcpConn, ok := conn.(*net.TCPConn); ok {
+										tcpConn.SetLinger(0)
+									}
+									conn.Close()
+									middlewareLog.Debugw("response truncated due to timeout",
+										"status", statusCode,
+										"bytesWritten", tw.bytesWritten)
+								}
+							}
+						}
+
+						// Signal timeout to potentially waiting handler
+						close(timeoutChan)
+					}
+					tw.mu.Unlock()
+					return
+
+				case <-handlerDone:
+					// Handler completed, stop timer and exit
+					tw.timer.Stop()
+					return
+				}
+			}
+		}()
+
+		// Run handler in a goroutine so we can interrupt it on timeout
+		go func() {
+			defer func() {
+				tw.mu.Lock()
+				tw.handlerComplete = true
+				tw.mu.Unlock()
+				close(handlerDone)
+			}()
+			handler.ServeHTTP(tw, r)
```

Comment:
The `defer func()  { }` seem unnecessary.
```suggestion
			handler.ServeHTTP(tw, r)
			tw.mu.Lock()
			tw.handlerComplete = true
			tw.mu.Unlock()
			close(handlerDone)
```

Reply:
@gammazero unsure if i got it right, but we may want to keep that `defer` due to panics:
- gateway code is able to recover from panics - see `panicHandler` in `./gateway/handler.go` - we do this to not crash the daemon if someone asks for a funny CID that triggers a bug
- so if `handler.ServeHTTP(tw, r)` ever panics, we still need to close the channel and set `handlerComplete`, otherwise the timeout goroutine will wait, no?
  - so iirc  the `defer` IS necessary for safety. However, we could make it simpler and only have `defer close(handlerDone)` before calling `handler.ServeHTTP` – I pushed this fix in a68079b65cbc9e496f7a43d65bae0db092f24515

Reply:
OK, no change needed then. Doing this  in a defer func will ensure it gets done even if there is a panic.